src/Catalog/PackageDeprecationItem.cs

```diff
@@ -0,0 +1,70 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using NuGet.Versioning;
+
+namespace NuGet.Services.Metadata.Catalog
+{
+    public class PackageDeprecationItem
+    {
+        public PackageDeprecationItem(
+            IReadOnlyCollection<string> reasons,
+            string message = null,
+            PackageDeprecationAlternatePackageItem alternatePackage = null)
+        {
+            if (reasons == null)
+            {
+                throw new ArgumentNullException(nameof(reasons));
+            }
+
+            if (!reasons.Any())
+            {
+                throw new ArgumentException(nameof(reasons));
+            }
+
+            Reasons = reasons;
+            Message = message;
+            AlternatePackage = alternatePackage;
+        }
+
+        public IReadOnlyCollection<string> Reasons { get; }
+        public string Message { get; }
+        public PackageDeprecationAlternatePackageItem AlternatePackage { get; }
+
+        public class PackageDeprecationAlternatePackageItem
```

Comment:
Nesting classes is not necessary here. Makes things more verbose than is necessary.

Reply:
I wanted to better enforce that the only possible states for a `PackageDeprecationItem` are
- no alternate package
- alternate package with both ID and version specified

Now that I think about it, it would be easier to just enforce those things in the constructor.

Reply:
I don't understand how nesting classes and validation are related.

Reply:
With a nested class, callers can only provide either null or a `PackageDeprecationAlternatePackageItem`. This enforces that they can't provide an invalid combination of IDs and versions to this constructor...but in retrospect it just moves that logic to the constructor or `PackageDeprecationAlternatePackageItem`. I removed the nested classes to simplify the code.